src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs

```diff
@@ -14180,5 +14180,189 @@ .maxstack  1
             verifier.VerifyIL("NativeInts.Checked4", expectedCheckedIL);
             verifier.VerifyIL("NativeInts.Checked5", expectedCheckedIL);
         }
+                             
+        [WorkItem(47887, "https://github.com/dotnet/roslyn/issues/47887")]
+        [Fact]
+        public void ArrayAccessNInt()
+        {
+            string source =
+@"class C
+{
+    static void Main()
+    {
+        int r = F(new int[] { 1 }, 0);
+        System.Console.WriteLine(r);
+    }
+    static int F(int[] a, nint i)
```

Comment:
Please also add tests for IntPtr/UintPtr types.

Reply:
There is an identity conversion from the `IntPtr`/`UIntPtr` to `nint/`nuint`, respectively.

The array indexer is functionally similar to a method `T this[nint index]` and therefore `IntPtr`/`UIntPtr` should be allowed, correct?

Reply:
>There is an identity conversion from the IntPtr/UIntPtr to nint/nuint, respectively.

>The array indexer is functionally similar to a methodT this[nint index]and thereforeIntPtr/UIntPtr` should be allowed, correct?

The fact that there is an identity conversion between the types doesn't mean that there is no difference in how the types are treated. It looks like, given the current language specification, indexing with IntPtr instances is not allowed. I am not sure if it is worth supporting this scenario. Unless we propose a change for the language specification and get it approved, I expect this scenario to keep failing.


---
In reply to: [568065999](https://github.com/dotnet/roslyn/pull/50910#discussion_r568065999) [](ancestors = 568065999)

Reply:
Is there an existing mechanism to say "allow nint but not IntPtr" here?

As it is, this will go through: `http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Binder_Expressions.cs,7434` which will have `SpecialType.System_IntPtr` where `IsNativeIntegerType` will be `false` as that's what `GetSpecialType` returns. I don't see an immediately obvious overload to return a variant for `nint`/`nuint` and that would likely end up passing the `ClassifyImplicitConversionFromExpression` check anyways.

So, is the appropriate thing to just explicitly check and handle (pseudo-code) `targetType is IntPtr` and `!expr.Type.IsNativeInteger` in `TryImplicitConversionToArrayIndex`?

Reply:
> So, is the appropriate thing to just explicitly check and handle (pseudo-code) targetType is IntPtr and !expr.Type.IsNativeInteger in TryImplicitConversionToArrayIndex?

I suggest to stop "messing" with the binding. The stated goal is to change what IL is generated, so suggest investigating if we can simply change how ldelem and other relevant instructions are emitted.

---
In reply to: [568085620](https://github.com/dotnet/roslyn/pull/50910#discussion_r568085620) [](ancestors = 568085620)

Reply:
See ```CodeGenerator.EmitArrayElementLoad``` and the like. In there we probably could detect that we are dealing with a "noop converted" IntPtr/UintPtr as an szarray index and skip the conversion


---
In reply to: [568090939](https://github.com/dotnet/roslyn/pull/50910#discussion_r568090939) [](ancestors = 568090939,568085620)

Reply:
Are there any limitations to adjusting `Emit` here?

What I've prototyped locally is to update: [EmitArrayIndices](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/CodeGen/EmitAddress.cs,380)
to basically the following, which results in ideal or nearly ideal codegen and preserves possible side-effects where that is relevant:

```csharp
private void EmitArrayIndices(ImmutableArray<BoundExpression> indices)
{
    for (int i = 0; i < indices.Length; ++i)
    {
        BoundExpression index = indices[i];
        TreatLongsAsNative(index, used: true);
    }
}

private void TreatLongsAsNative(BoundExpression index, bool used)
{
    Microsoft.Cci.PrimitiveTypeCode tc = index.Type.PrimitiveTypeCode;

    var code = ILOpCode.Nop;

    if (tc == Microsoft.Cci.PrimitiveTypeCode.Int64)
    {
        code = ILOpCode.Conv_ovf_i;
    }
    else if (tc == Microsoft.Cci.PrimitiveTypeCode.UInt64)
    {
        code = ILOpCode.Conv_ovf_i_un;
    }

    if (code != ILOpCode.Nop)
    {
        var constantValue = index.ConstantValue;

        if ((constantValue != null) && constantValue.IsIntegral)
        {
            ConstantValueTypeDiscriminator discriminator = constantValue.Discriminator;

            switch (discriminator)
            {
                case ConstantValueTypeDiscriminator.SByte:
                case ConstantValueTypeDiscriminator.Byte:
                case ConstantValueTypeDiscriminator.Int16:
                case ConstantValueTypeDiscriminator.UInt16:
                case ConstantValueTypeDiscriminator.Int32:
                    {
                        // The constant is smaller than (or potentially equal, in the case of
                        // System_Int32) native integer so we can ignore the cast to long/ulong
                        // and instead emit a direct unchecked cast to native int. 

                        code = ILOpCode.Conv_i;
                        break;
                    }

                case ConstantValueTypeDiscriminator.UInt32:
                case ConstantValueTypeDiscriminator.NUInt:
                    {
                        // The constant is unsigned and may be equal in length or smaller than native integer
                        // while ldelem only officially supports int32 or native int. If the constant is
                        // less than or equal to int.MaxValue, we can just emit a direct unchecked cast to
                        // native int; otherwise we need a checked cast to native int.

                        if (constantValue.UInt32Value <= int.MaxValue)
                        {
                            code = ILOpCode.Conv_i;
                        }
                        else
                        {
                            code = ILOpCode.Conv_ovf_i_un;
                        }
                        break;
                    }

                case ConstantValueTypeDiscriminator.Int64:
                    {
                        // The constant is signed but may be equal in length or larger than native integer so
                        // we can emit a unchecked cast if the value is an int32 and a checked cast otherwise.

                        if ((constantValue.Int64Value >= int.MinValue) && (constantValue.Int64Value <= int.MaxValue))
                        {
                            // This will end up with a IL sequence such as:
                            //   ldc.i4.1
                            //   conv.i8
                            //   conv.i
                            //
                            // This isn't "ideal" due to the intermediate conv.i8, but it's still
                            // beter than inserting an unnecessary overflow check.

                            code = ILOpCode.Conv_i;
                        }
                        else
                        {
                            code = ILOpCode.Conv_ovf_i;
                        }
                        break;
                    }

                case ConstantValueTypeDiscriminator.UInt64:
                    {
                        // The constant is unsigned but may be equal in length or larger than native integer so
                        // we can emit a unchecked cast if the value is an int32 and a checked cast otherwise.

                        if (constantValue.UInt64Value <= int.MaxValue)
                        {
                            // This will end up with a IL sequence such as:
                            //   ldc.i4.1
                            //   conv.u8
                            //   conv.i
                            //
                            // This isn't "ideal" due to the intermediate conv.u8, but it's still
                            // beter than inserting an unnecessary overflow check.

                            code = ILOpCode.Conv_i;
                        }
                        else
                        {
                            code = ILOpCode.Conv_ovf_i_un;
                        }
                        break;
                    }

                case ConstantValueTypeDiscriminator.NInt:
                    {
                        // The constant is a native int and of the right sign for ldelem, so
                        // we can just avoid any conversion being emitted.

                        code = ILOpCode.Nop;
                        break;
                    }
            }
        }
        else if (index is BoundConversion conversion)
        {
            BoundExpression operand = conversion.Operand;

            switch (operand.Type.SpecialType)
            {
                case SpecialType.System_SByte:
                case SpecialType.System_Byte:
                case SpecialType.System_Int16:
                case SpecialType.System_UInt16:
                case SpecialType.System_Int32:
                    {
                        // These types are smaller than (or potentially equal, in the case of
                        // System_Int32) native integer so we can ignore the cast to long/ulong
                        // and instead emit a direct unchecked cast to native int.

                        code = ILOpCode.Conv_i;
                        index = operand;
                        break;
                    }

                case SpecialType.System_UInt32:
                case SpecialType.System_UIntPtr:
                    {
                        // These types are unsigned and may be equal in length or smaller than native integer
                        // while ldelem only officially supports int32 or native int, so we need to use an
                        // overflow conversion to nint but can again bypass the original conversion to long/ulong.

                        code = ILOpCode.Conv_ovf_i_un;
                        index = operand;
                        break;
                    }

                case SpecialType.System_IntPtr:
                    {
                        // The type is native int and of the right sign for ldelem, so we can just
                        // avoid any conversion being emitted.

                        code = ILOpCode.Nop;
                        index = operand;
                        break;
                    }
            }
        }
    }

    EmitExpression(index, used: true);

    if (code != ILOpCode.Nop)
    {
        _builder.EmitOpCode(code);
    }
}
```

Reply:
I pushed changes that do something similar to what I said above and added tests covering the various patterns for both array access and array creation

Reply:
For what it's worth, I think that handling this in emit is actually quite a bit more complicated, and potentially error prone, then having this handled in binding and letting it flow through "naturally".

In binding, I had determined single dimensional zero-based arrays can be checked for via `ArrayTypeSymbol.HasDefaultSizesAndLowerBounds`. It also avoids the issue with the constant nodes being `int64` when we want them to be `nint` or `nuint`, makes it clearer when overflow is or is not important, and I think makes the language slightly more cohesive to have `nint`/nuint` properly supported (considering this is the "correct" type for indexing anyways; since memory and references are naturally `nint`/`nuint` sized).

Reply:
> For what it's worth, I think that handling this in emit is actually quite a bit more complicated, and potentially error prone, then having this handled in binding and letting it flow through "naturally".

The nature of the changes that were made in binding resulted in observable changes in behavior of APIs that compiler implements and likely resulted in unintended binding behavior changes. Though it might feel simpler to adjust binding behavior in this way in order to achieve desired IL gen outcome, that would cause compiler to behave not according to the language specification in an observable way. This would be the wrong thing to do.

---
In reply to: [568417406](https://github.com/dotnet/roslyn/pull/50910#discussion_r568417406) [](ancestors = 568417406)

Reply:
> that would cause compiler to behave not according to the language specification in an observable way. This would be the wrong thing to do.

Would it then be beneficial, long term, to get this as an approved change to the language spec? I don't mind which approach is done, or even if we do one today and another as a long term goal.
Just want to understand the correct approach and to understand the benefits/drawbacks of each.

Reply:
> Would it then be beneficial, long term, to get this as an approved change to the language spec?

What updates would we make though? Today expressions of type `nint` and `nuint` are legal according to the C# spec because they have implicit conversions to `long` and `ulong` respectively. Hence they fall into the following rule:

> each expression must be of type int, uint, long, ulong, or must be implicitly convertible to one or more of these types

This bug is just about how array indexes using these types are emitted as IL, essentially we could be a lot more efficient here. Hence that is what the PR should be focusing on changing. I'm definitely supportive of that change given that it's at worst execution neutral and reduces the size of emitted IL (which is a current goal of the runtime team). But don't see a need at the moment to bring language changes with it.

I can see some value over the long term in adjusting both the language space and ECMA 335 to allow `nint` and `nuint` as array indexes. Essentially make the language spec line up with the ideal way that the runtime wants to see the IL. That's a convo that would need to start on csharplang though.

Reply:
> What updates would we make though

Right, it would change it to explicitly allow `nint` and `nuint`. Likewise, it would explicitly allow `nint op pointer`, `nuint op pointer`, `pointer op nint`, and `pointer op nuint` for all the same places these are allowed with `int`, `uint`, `long`, and `ulong`.

> That's a convo that would need to start on csharplang though.

I'll see if I can get something up this weekend proposing this. I don't think it would block optimizing the codegen in the current PR either way.

Reply:
> Would it then be beneficial, long term, to get this as an approved change to the language spec? I don't mind which approach is done, or even if we do one today and another as a long term goal.
Just want to understand the correct approach and to understand the benefits/drawbacks of each.

At this point, I will support a change to language specification that will allow us to avoid performing "unnecessary" conversions for nint/nuint indexes as long as it is worded in such a way that it doesn't enable any new scenarios or conversions. Something like: "for szarrays each expression must be of type int, uint, long, ulong, nint, nuint or must be implicitly convertible to int, uint, long or ulong? Anything other than that would require more careful consideration and likely more complicated change in the compiler (language version check, etc.). Let us know which path you would like to pursue. It probably doesn't make sense going through the trouble of changing emit layer and then moving changes to binding.

---
In reply to: [568894825](https://github.com/dotnet/roslyn/pull/50910#discussion_r568894825) [](ancestors = 568894825)

Reply:
> I don't think it would block optimizing the codegen in the current PR either way.

I do not believe there is urgency to do something here. I'd prefer us to avoid unnecessary churn in the code base. Let's decide what, if anything, we want to change in the language specification. 

---
In reply to: [569501237](https://github.com/dotnet/roslyn/pull/50910#discussion_r569501237) [](ancestors = 569501237)

Reply:
> Let us know which path you would like to pursue

I believe that properly supporting `nint`/`nuint` would be better for the language and ecosystem long term, so I'll close this PR and work on getting the language proposal up instead.

> It probably doesn't make sense going through the trouble of changing emit layer and then moving changes to binding.

I think I generally agree.

After having prototyped the emitter changes, the main benefit would be to catch cases where the user did something such as: `array[(long)nintOrSmallerValue]`
However, if `nint` is properly supported then users have an explicit way to get the "better" (no overflow checking) behavior instead.

Additionally, even with the emit fix right now, there is still some general wonkiness when dealing with things like `nint constants`.
Also, adding support to ensure all the `long->nint` and `ulong->nint` scenarios generate good code ends up being quite complicated and involves correctly tracking cases that could have been `nint`, which makes the logic not trivial to follow/reason about.

> Something like: "for szarrays each expression must be of type int, uint, long, ulong, nint, nuint or must be implicitly convertible to int, uint, long or ulong

I think that's reasonable. Given the current setup;
* `byte`, `sbyte`, `short`, `ushort`, and `int` already become: `int`.
* `uint`, `nint`, and `long` become `long`
* `nuint` and `ulong` become `ulong`

This would effectively just be changing it so that:
* `byte`, `sbyte`, `short`, `ushort`, and `int` already become: `int`.
* `nint` becomes `nint`
* `nuint` becomes `nuint`
* `uint`, and `long` become `long`
* `ulong` becomes `ulong`

Given the implicit conversions `nint`/`nuint` support today and subtracting what is already covered by `int`; the only changes from allowing them would be `IntPtr`->`nint` (undesirable), `UIntPtr`->`nuint` (undesirable), and `uint->nuint` (possibly desirable, but users can always explicitly convert)

Reply:
I've opened https://github.com/dotnet/csharplang/issues/4385 and believe I have found the relevant spec locations that would need to be updated to support `nint` and `nuint` without also supporting implicit conversions to `nint`/`nuint`